tensorboard/webapp/plugins/npmi/data_source/npmi_data_source.ts

```diff
@@ -40,6 +54,61 @@ export class NpmiHttpServerDataSource implements NpmiDataSource {
       this.fetchAnnotations(),
       this.fetchMetrics(),
       this.fetchValues()
+    ).pipe(
+      map(([annotations, metrics, values]) => {
+        let annotationData: AnnotationDataListing = {};
+        for (let run of Object.keys(annotations)) {
+          for (let annotationIndex in annotations[run]) {
+            let annotation = annotations[run][annotationIndex];
+            let dataElements: ValueData[] = [];
+            for (let metricIndex in metrics[run]) {
+              let metric = metrics[run][metricIndex];
+              let dataElement = dataElements.find(
+                (element) =>
+                  element.metric === metric_type.stripMetricString(metric)
+              );
+              if (!dataElement) {
+                dataElement = {
+                  nPMIValue: null,
+                  countValue: null,
+                  annotation: annotation,
+                  metric: metric_type.stripMetricString(metric),
+                  run: run,
+                };
+                dataElements.push(dataElement);
+              }
+              if (metric_type.metricIsMetricCount(metric)) {
+                dataElement.countValue =
+                  values[run][annotationIndex][metricIndex];
+              } else if (metric_type.metricIsNpmi(metric)) {
+                dataElement.nPMIValue =
+                  values[run][annotationIndex][metricIndex];
+              }
+            }
+            if (annotationData[annotation]) {
+              annotationData = {
+                ...annotationData,
+                [annotation]: annotationData[annotation].concat(dataElements),
+              };
+            } else {
+              annotationData = {
+                ...annotationData,
+                [annotation]: dataElements,
+              };
+            }
+          }
+        }
+        return {annotationData, metrics};
+      }),
+      catchError((error) => {
+        if (error instanceof HttpErrorResponse && error.status === 400) {
```

Comment:
Is there any significance to 400 exactly? Or do you care about anything in between `400 <= code < 500`?

Reply:
I think I adopted that from here: `google3/third_party/tensorboard/google/ui/app/runs/data_source/runs_data_source.ts`. But for my plugin, I agree that this does not really make sense. So changing it to your proposal.